students/Chao/project_fpy/fpy.py

```diff
@@ -0,0 +1,253 @@
+import getpass
+import ipaddress
+
+class LTM:
+    """ Main Class for Local Traffic Manager Module """
+    def __init__(self, mgmt):
+        self.mgmt = mgmt
+
+class Pool(LTM):
+    """ Pool class for retrieving pool status """
+
+    def get_pool_state(self, p_name, p_partition='Common'):
+        """ Function to retrieve current pool availability status """
+        pool_stat = self.mgmt.tm.ltm.pools.pool.load(name=p_name, partition=p_partition)
+        stats = Stats(pool_stat.stats.load())
+        return stats.stat.status_availabilityState['description']
+
+    def get_pool(self):
+        """ Pool full status function """
+        pools = self.mgmt.tm.ltm.pools.get_collection()
+        print("\n==================== Pool Status ========================")
+        print("|       Pool Name      |    Partition    | Availability |")
+        for pool in pools:
+            pstat = self.get_pool_state(pool.name, pool.partition)
+            print("| {:20} | {:15} | {:12} |".format(pool.name, pool.partition, pstat))
+        print("=========================================================")
+
+class Virtual(LTM):
+    """ Virtual class for retrieving virtual server status or modifying config """
+
+    def get_virtual_state(self, v_name, v_partition='Common'):
+        """ Function to retrieve current virtual server availability status """
+        virtual_stat = self.mgmt.tm.ltm.virtuals.virtual.load(name=v_name, partition=v_partition)
+        stats = Stats(virtual_stat.stats.load())
+        return stats.stat.status_availabilityState['description']
+
+    def get_virtual(self):
+        """ Virtual Server full status function """
+        virtuals = self.mgmt.tm.ltm.virtuals.get_collection()
+        print("\n=================== Virtual Status ======================")
+        print("|  Virtual Server Name |    Partition    | Availability |")
+        for virtual in virtuals:
+            vstat = self.get_virtual_state(virtual.name, virtual.partition)
+            print("| {:20} | {:15} | {:12} |".format(virtual.name, virtual.partition, vstat))
+        print("=========================================================")
+
+    def apply_rule(self, myrule='fpy_maintenace'):
+        """ iRule apply function """
+        virtuals = self.mgmt.tm.ltm.virtuals.get_collection()
+        for virtual in virtuals:
+            # set full iRule path
+            rule_path = '/{}/{}'.format(virtual.partition, myrule)
+            vrule = self.mgmt.tm.ltm.virtuals.virtual.load(name=virtual.name, partition=virtual.partition)
+            try:
+                if rule_path in vrule.rules:
+                    print("Rule {} already exists in {}".format(myrule, virtual.name))
+                else:
+                    # Insert iRule at the front of the list
+                    vrule.rules.insert(0, rule_path)
+                    vrule.update()
+                    print("Rule {} added to {}".format(myrule, virtual.name))
```

Comment:
I think this would be better:
```
if rule_path in vrule.rules:
    print("Rule {} already exists in {}".format(myrule, virtual.name))```
    return
try:
    # Insert iRule at the front of the list
    vrule.rules.insert(0, rule_path)
    vrule.update()
except iControlUnexpectedHTTPError:
    print("Unable to apply rule {} to {}. HTTP profile required.".format(myrule, virtual.name))
else:
    print("Rule {} added to {}".format(myrule, virtual.name))
```
It is pretty minor, really, but having conditionals in try/except blocks can lead to confusion and occasionally unexpected errors.